# Practice Case MongoDB Day 1

### Install Library
First of all we need to install the library that will be used for this practice

In [122]:
!pip install pymongo
!pip install pymongo[srv]
!pip install pprint
!pip install iPython

# Instruction 1

Make **new collection** with 'clean_movies' name that same as **movies** collection in **sample_mfilx** database using collection **movies_initial** <br>

## Connect Cluster PyMongo
After the library successfully installed, we need to load the data by take the data from Cluster in MongoDB with connection link that served by MongoDB.<br>
<br>
Based of the intruction, we need to load the data from **two cluster** because the **newest cluster doesnt include movies collection**.

### New Cluster

In [123]:
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

client = MongoClient("mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority")

print(client)

MongoClient(host=['cluster0-shard-00-00-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-01-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-02-miqju.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


### Old Cluster

In [124]:
clients = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true& w=majority")

print(clients)

MongoClient(host=['cluster0-shard-00-02-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-00-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-01-nnbxe.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


/usr/local/lib/python3.6/dist-packages/pymongo/common.py:756: UserWarning: Unknown option  w
  warnings.warn(str(exc))


## Database in Cluster

In [125]:
print(client.list_database_names())

['sample_mflix', 'admin', 'local']


In [217]:
print(clients.list_database_names())

['sample_mflix', 'admin', 'local']


There was database that contain **only** **sample_mflix** that we want use to running the instruction.

## Load Database
After the cluster made, we need to load the database that we want to use to solve the instruction.

In [0]:
db=client['sample_mflix'] 

In [127]:
print(db.list_collection_names())

['clean_movies_Putrisqiana', 'clean_movies_tara', 'clean_movies_bunga', 'clean_movies_zumar', 'movies_initial', 'clean_movies_putrisqiana', 'clean_movies_firdaus', 'clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_faizah']


In [0]:
database=clients['sample_mflix']

In [129]:
print(database.list_collection_names())

['clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_devita', 'movies_initial', 'clean_movies_frans', 'clean_movies_tara', 'clean_movies_tunggal', 'clean_movies', 'users', 'clean_movies_nurrimah', 'movies_testtopik', 'movies_scratch', 'clean_movies_topik', 'movies', 'clean_movies_afifah', 'clean_movies_andreas', 'clean_movies_indra_1']


## Load Collection
After the database is loaded, we need to choose the collection that we want to use to solve the instruction by loading them in a variable.

### Movies Initial

In [0]:
col=db.movies_initial

In [218]:
# Show keys in movies_initial
x = col.find_one()
print(x.keys())

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])


### Movies Collection

In [0]:
collection=database.movies

In [133]:
# Show keys in choosen collection
x = collection.find_one()
print(x.keys())

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])


In [0]:
import pandas as pd

In [135]:
pd.DataFrame(list(col.find().limit(1)))

_id awards        cast  ...   type writer  year
0  5daffb1b986c4adedb754af9         Carmencita  ...  movie         1894

[1 rows x 22 columns]

In [136]:
pd.DataFrame(list(collection.find().limit(1)))

_id  ...  year
0  573a1390f29313caabcd5a93  ...  1916

[1 rows x 19 columns]

Because the intruction is need to **use movies as the collection in new collection**, in the movies collection there was dictionary like:
* _id
* awards
* cast
* countries
* directors
* fullplot
* genres
* imdb
* languages
* lastupdated
* metacritic
* num_mflix_comments
* plot
* poster
* rated
* released
* runtime
* title
* tomatoes
* type
* writers
* year

The dictionary of new collection that taken by **movies_initial collection** should be the same like the in the **movies collection**


## Make a new collection

In [203]:
pipeline = [
    {
        '$project': {
            'title': 1,
            'year': 1,
            'plot': 1,
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'directors': {'$split': ["$director", ", "]},
            'cast': {'$split': ["$cast", ", "]},
            'writers': {'$split': ["$writer", ", "]},
            'languages': {'$split': ["$language", ", "]},
            'genres': {'$split': ["$genre", ", "]},
            'countries': {'$split': ["$country", ", "]},
            'fullplot': "$fullplot",
            'released': { '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] },
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'num_mflix_comments': {'$sum':1},
            'poster': 1,
            'rated': "$rating",
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'type': 1,
            'lastupdated': "$lastupdated",
            'tomatoes':{
                'lastupdated':'$lastupdated',
                'viewer': {'meter':'', 'numReviews':'', 'rating':''},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''
                
            }
        }
    },
    
    #matching with imdb_list_2 that has been created before but need to sorted like imdb format
    { "$match": {"imdb": {"$in" : [dict(sorted(imdb_list_2[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}}
    
    #New Collection name is 'clean_movies_putrisqiana'
    ,{'$out': 'clean_movies_putrisqiana'}
]

clear_output()
pprint.pprint(list(client.sample_mflix.movies_initial.aggregate(pipeline)))

[]


The new collection must to be same as movie collection, so the field and values need to be change in this code. <br>
<br>
We need the last line of the code to take the data and aggreate the pipeline to make the new collection.
<br>
<br>
Field that same as movie collection without changes is :
1. Title
2. Year
3. plot
4. poster
5. type.
6. fullplot
7. lastupdated

Field that change the name of key is and need to split the value is:
1. director --> directors (both)
2. writer --> writers (both)
3. language --> languages (both)
4. genre --> genres (both)
5. country --> countries (both)
6. rated --> ratings (change name only)
7. cast (split only)

Field that the value has been fill if it consist, and have no value if it not consist :
1. metacritic
2. runtime

field that need to concat : released

field that made into one of several : imdb. consist of imdbID, imdbRating and ImdbVotes.

field that the values is always 1 : num_mflix_comment

field that the value is need to be change in that conditions
1. awards
2. tomatoes






# Data Validation

# Instruction 2

## Count the Data

In [137]:
len(list(col.find({})))

46014

In [138]:
len(list(collection.find({})))

23541

The amount of document in movies_initial **col** *is different* with the amount of data in movies **collection**<br>
<br>
So wee need to make them same.

## Matching data

### 1st Looping
Make all the document in collection become an array format from the IMDB field, because the IMDB field in movies collection consist of three field in movies_initial collection.

In [0]:
#1st looping to get imdb_list that are not in array format
imdb_list = []
for imdb in list(collection.find({}, {'_id':0, 'imdb':1})):
  imdb_list.append(imdb.values())

In [147]:
imdb_list[0]

dict_values([{'rating': 6.3, 'votes': 162, 'id': 6517}])

### 2nd Looping
Get list of array format

In [0]:
#2nd looping to get list of array format
imdb_list_2 = []
for i in range(len(imdb_list)):
  imdb_list_2.append(list(imdb_list[i])[0])

In [149]:
#Check is the looping work or not
imdb_list_2[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

# Instruction 2.2.
**Amount of document** in clean_movies_putrisqiana collection and movie_collection are same.

In [0]:
db=client['sample_mflix'] 
colection = db.clean_movies_putrisqiana

In [178]:
len(list(colection.find()))

23539

In [142]:
len(list(collection.find()))

23541

There was 2 different document in **clean_movies_putrisqiana collection** and in **movies collection**

# Instruction 2.1.
All **document** in clean_movies and movie is same

In [0]:
putrisqianacollection = list(db['clean_movies_putrisqiana'].aggregate([{'$sortByCount':'$imdb'}]))
moviescollection = list(database['movies'].aggregate([{'$sortByCount':'$imdb'}]))

In [183]:
for doc1 in moviescollection:
  if doc1 not in putrisqianacollection:
    print(doc1)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


The different is on imdb with **'id' : 5**

In [185]:
#imdb id 5 di clean_movies_putrisqiana
list(db['clean_movies_putrisqiana'].find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'languages': [''],
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'poster': '',
  'rated': 'UNRATED',
  'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
  'runtime': '1',

In [188]:
#imdb id 5 di movies
list(database['movies'].find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5dafd0d3d75e5fa8b23dae21'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short', 'Long'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene II',
  'to

The different on id:5 in these two collection is **genres** keys which is in **clean_movies_putrisqiana collection** the values is only ['Short]', but in **movies collection** the value is ['Long', 'Short']

 # Instruction 2.3.
 **all document in clean_movies_putrisqiana collection and movies collection are same** <br>

In [204]:
# how to show all keys MOVIES in one vertical list
mine = [list(doc.keys()) for doc in colection.find({})]
mine=set([a for b in mine for a in b])
mine

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [205]:
movie = [list(doc.keys()) for doc in collection.find({})]
movie=set([c for d in movie for c in d])
movie

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [0]:
movies_keys = list(database['movies'].find()[4].keys())
putrisqiana_keys = list(db['clean_movies_putrisqiana'].find_one().keys())

In [0]:
# Checking different
for key in putrisqiana_keys:
  if key not in movies_keys:
    print(key)

 because in the '# Checking different' code, there was no keys that appear in the result.

In [206]:
len(mine)

22

In [207]:
len(movie)

22

**amount of document in clean_movies_putrisqiana collection and movies collection are same**. <br>
There was 22 document in both of collection.

# Instruction 2.4.
All value in clean_movies_putrisqiana collection has the same value and same order with movies collection and the same order.


In [215]:
#Chech one of document in movies randomly
database['movies'].find()[17]

{'_id': ObjectId('573a1390f29313caabcd50e5'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Winsor McCay', 'George McManus', 'Roy L. McCardell'],
 'countries': ['USA'],
 'directors': ['Winsor McCay'],
 'fullplot': 'Winsor Z. McCay bets another cartoonist that he can animate a dinosaur. So he draws a big friendly herbivore called Gertie. Then he get into his own picture. Gertie walks through the picture, eats a tree, meets her creator, and takes him carefully on her back for a ride.',
 'genres': ['Animation', 'Short', 'Comedy'],
 'imdb': {'id': 4008, 'rating': 7.3, 'votes': 1837},
 'languages': ['English'],
 'lastupdated': '2015-08-18 01:03:15.313000000',
 'num_mflix_comments': 1,
 'plot': 'The cartoonist, Winsor McCay, brings the Dinosaurus back to life in the figure of his latest creation, Gertie the Dinosaur.',
 'poster': 'https://m.media-amazon.com/images/M/MV5BMTQxNzI4ODQ3NF5BMl5BanBnXkFtZTgwNzY5NzMwMjE@._V1_SY1000_SX677_AL_.jpg',
 'released': datetime.date

In [216]:
#Then match it with the document in 'clean_movies_putrisqiana' with 'Gertie the Dinosaur'
list(db['clean_movies_putrisqiana'].find({'title': 'Gertie the Dinosaur'}))

[{'_id': ObjectId('5daffb1b986c4adedb754b50'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Winsor McCay', 'George McManus', 'Roy L. McCardell'],
  'countries': ['USA'],
  'directors': ['Winsor McCay'],
  'fullplot': 'Winsor Z. McCay bets another cartoonist that he can animate a dinosaur. So he draws a big friendly herbivore called Gertie. Then he get into his own picture. Gertie walks through the picture, eats a tree, meets her creator, and takes him carefully on her back for a ride.',
  'genres': ['Animation', 'Short', 'Comedy'],
  'imdb': {'id': 4008, 'rating': 7.3, 'votes': 1837},
  'languages': ['English'],
  'lastupdated': '2015-08-18 01:03:15.313000000',
  'num_mflix_comments': 1,
  'plot': 'The cartoonist, Winsor McCay, brings the Dinosaurus back to life in the figure of his latest creation, Gertie the Dinosaur.',
  'poster': 'http://ia.media-imdb.com/images/M/MV5BMTQxNzI4ODQ3NF5BMl5BanBnXkFtZTgwNzY5NzMwMjE@._V1_SX300.jpg',
  'rated': '',
  'rel

# Conclusion

* The new collection is successfully made with name 'clean_movies_putrisqiana'. The collection is made by taking the data from movies_initial collection in sample_mflix database.But, it need  to same as movies collection.

* Then, amount of document from both collection is different.<br> 
<br>**clean_movies_putrisqiana** contain of 23539 document. **movies** collection contain 23541 document<br>
<br> Because there was 2 document that cant change to be same.
The document is on id:5 in movies collection, genres field consist of multiple values but in clean_movies_putrisqiana genres field consist only one values.

* **Amount of field in both collection is same 22 fields** with the same names that consist of _id, awards, cast, countries, directors, fullplot, genres, imdb, languages, lastupdated, metacritic, num_mflix_comments, plot, poster, rated, released, runtime, title, tomatoes, type, writers, year.